handlers.py

```diff
@@ -48,6 +49,34 @@ async def reply(self, content):
         self.write(content)
         await self.finish()
 
+    def prepare(self):
+        ip_address = self.request.headers.get("X-Real-IP") or \
+                     self.request.headers.get("X-Forwarded-For") or \
+                     self.request.remote_ip
+        print(ip_address)
+
+        #if self.request_is_limited(key=ip_address, limit=self.application.limit, period=self.application.period):
+        #    raise tornado.web.HTTPError(429)
+
+    def _expire_requests(self, now: datetime, period: timedelta, usage, key):
+        """ remove all requests before expiration time"""
+        expiration_time = now - period
+        i = 0
+        for i, usage in enumerate(usage[key]):
+            if usage > expiration_time:
+                break
+        if i > 0:
+            self.usage[key] = self.usage[key][i:]
+
+    def request_is_limited(self, key: str, limit: int, period: timedelta):
+        now = datetime.utcnow()
+        self._expire_requests(now, period, self.usage, key)
+        limit_reached = len(self.usage[key]) > limit
+        if limit_reached:
+            return True
+        self.usage[key].append(now)
```

Comment:
@ishefi I'm not sure. if we keep it like this a person could poll to check if the usage has expired, and will be able to figure out the limit fairly easily. I can fix this be adding the request even if the limit has passed, but that exposes us that a person will ddos the system (and fill lots of values in self.usage[key]). I can also save only the last X most recent results (with X=limit). what do you think?

Reply:
1. why not use a counter?
2. why not use [`@functools.lru_cache`](https://docs.python.org/3/library/functools.html) somehow?

Reply:
1. ok to use a counter, but it won't be a sliding window (which might be fine). that is we won't be able to enforce 10 requests every 20 seconds, but we would be able to enforce 10 requests in every interval of 20 minutes.
2. lru cache can work... just wanted to know what you think (whether we should count limited requests as requests that we count or not)